src/serialization.rs

```diff
@@ -0,0 +1,693 @@
+            slice
+                .as_rust_type_mut_uninit()
+                .write(CSliceOwned::default());
+            result::Z_EIO
+        }
+    }
+}
+
+/// Serializes a string.
+#[no_mangle]
+pub extern "C" fn ze_serializer_serialize_string(
+    this: &mut ze_loaned_serializer_t,
+    str: &z_loaned_string_t,
+) {
+    let cslice = str.as_rust_type_ref().slice();
+    this.as_rust_type_mut().serialize(cslice);
+}
+
+/// Serializes a null-terminated string.
+#[no_mangle]
+pub extern "C" fn ze_serializer_serialize_str(
+    this: &mut ze_loaned_serializer_t,
+    str: *const libc::c_char,
+) {
+    let slice = unsafe { from_raw_parts(str as *const u8, strlen(str)) };
+    this.as_rust_type_mut().serialize(slice);
+}
+
+/// Deserializes into a string.
+#[no_mangle]
+pub extern "C" fn ze_deserializer_deserialize_string(
+    this: &mut ze_deserializer_t,
+    str: &mut MaybeUninit<z_owned_string_t>,
+) -> z_result_t {
+    match this.as_rust_type_mut().deserialize::<String>() {
+        Ok(s) => {
+            str.as_rust_type_mut_uninit().write(s.into());
+            result::Z_OK
+        }
+        Err(e) => {
+            tracing::error!("Failed to deserialize the payload: {}", e);
+            str.as_rust_type_mut_uninit().write(CStringOwned::default());
+            result::Z_EIO
+        }
+    }
+}
+
+/// Initiates serialization of a sequence of multiple elements.
+/// @param this_: A serializer instance.
+/// @param len: Length of the sequence. Could be read during deserialization using `ze_deserializer_deserialize_sequence_begin`.
+#[no_mangle]
+pub extern "C" fn ze_serializer_serialize_sequence_begin(
+    this: &mut ze_loaned_serializer_t,
+    len: usize,
+) {
+    this.as_rust_type_mut().serialize(VarInt::<usize>(len));
+}
+
+/// Finalizes serialization of a sequence of multiple elements.
+#[no_mangle]
+#[allow(unused_variables)]
+pub extern "C" fn ze_serializer_serialize_sequence_end(this: &mut ze_loaned_serializer_t) {}
+
+/// Initiates deserialization of a sequence of multiple elements.
+/// @param this_: A serializer instance.
+/// @param len:  pointer where the length of the sequence (previously passed via `z_bytes_writer_serialize_sequence_begin`) will be written.
+/// @return 0 in case of success, negative error code otherwise.
+#[no_mangle]
+pub extern "C" fn ze_deserializer_deserialize_sequence_begin(
+    this: &mut ze_deserializer_t,
+    len: &mut usize,
+) -> z_result_t {
+    match this.as_rust_type_mut().deserialize::<VarInt<usize>>() {
+        Ok(l) => {
+            *len = l.0;
+            result::Z_OK
+        }
+        Err(e) => {
+            tracing::error!("Failed to read the sequence length: {}", e);
+            *len = 0;
+            result::Z_EIO
+        }
+    }
+}
+
+/// Finalizes deserialization of a sequence of multiple elements.
+#[no_mangle]
+#[allow(unused_variables)]
+pub extern "C" fn ze_deserializer_deserialize_sequence_end(
```

Comment:
I believe we can safely remove these sequence end functions. They do nothing at this moment and there is no guarantee at all that users will use them. Even if we add any optimzation based on these functions, this optimization will definitely break user's code without these functions.

If later we implement some optimized way of preparing / reading sequences which require this "end" function, it can be implemeted as completely new API, parallel to this one.

@wyfo even recommended to remove the "begin sequence" and just write the number, but I think it's excessive, the "begin sequence" as an alias which explicitly declares user's intention is useful. But end sequence should be removed I think

Reply:
done, renamed into ..._sequence_begin is renamed into  ..._sequence_length, ..._sequence_end is removed